### DISCUSSION TITLE
always opens file in browser on add regardless of settings

### DISCUSSION THREAD

**Discussion by ExpandingMan** [papis/papis#546] [Q&A]
I have the config
```
[settings]
default-library = papers
opentool = rifle
file-browser = ranger
use-cache = False
add-open = False

[tui]
editmode = vi


## SPECIFIC LIBRARIES ##

[papers]
dir = ~/Documents/papers
use-cache = False
```
But, despite this, papis still opens files in a browser (and for some reason not `rifle`) every time I add something.  Is there a way to prevent this?

**Comment by alexfikl** [comment]
What command are you running exactly? Is it `papis open` or `papis add`?

There's also a `browser` configuration option that may come into play there (see [here](https://papis.readthedocs.io/en/latest/configuration.html#tools-options)).

**Comment by ExpandingMan** [comment]
This is with `papis add`.  `papis open` works as I'd expect and opens the file only in the program I've specified (which is via `rifle` and does *NOT* use the browser).  That's part of what looks weird about the behavior of `papis add`, even if it did open it, I wouldn't expect it to be in the browser.

**Reply by alexfikl** [reply]
Hm, `papis add` should just call [`papis.utils.open_file`](https://github.com/papis/papis/blob/cf67c8f164ff4308d058e99c92e412ef5109609a/papis/utils.py#L163) which uses `opentool` and should complain pretty loudly if it doesn't succeed.

Can you share more exactly what command you're running? Maybe with some logs? I can't find anything in there that looks fishy and trying something like
```
papis add --from arxiv 'https://arxiv.org/abs/2305.00377'
```
seems to correctly open my PDF viewer with the downloaded file. I'm using `xdg-open`, not `rifle` though :confused:

**Reply by alexfikl** [reply]
One problem that comes to mind is that we save the downloaded file somewhere like `/tmp/tmp9m_swx1x`, without an extension, which might confuse `rifle` (?). I'll try to see if adding something there helps if we don't figure something out..

**Comment by ExpandingMan** [comment]
Ah, yes the lack of extension must confuse `rifle`, probably my config for it needs to be fixed.

However, even when, instead of `rifle` I use `zathura` directly, it still opens the file even when I set `add-open = False` in the config.

**Reply by alexfikl** [reply]
> Ah, yes the lack of extension must confuse `rifle`, probably my config for it needs to be fixed.

Most files also have some magic signature, e.g PDF files start with `%PDF-X.Y%` with some version `X.Y`, so if `rifle` doesn't look at that it won't be able to open it properly. At least the default config for `rifle` doesn't seem to work without the extension (installed it quickly).

I'll try to look if we can add some extensions to those files to help `rifle` out.

> However, even when, instead of `rifle` I use `zathura` directly, it still opens the file even when I set `add-open = False` in the config.

In the config, [`add-open = False`](https://github.com/papis/papis/blob/cf67c8f164ff4308d058e99c92e412ef5109609a/papis/defaults.py#L43) should be the default. I'm guessing what you're seeing is the files from an importer getting opened with a `Use this file? (from IMPORTER) (Y/n)` message before? 

Those should probably also respect `add-open = False` for consistency, so worth fixing :rocket:

**Comment by ExpandingMan** [comment]
> In the config, [add-open = False](https://github.com/papis/papis/blob/cf67c8f164ff4308d058e99c92e412ef5109609a/papis/defaults.py#L43) should be the default. I'm guessing what you're seeing is the files from an importer getting opened with a Use this file? (from IMPORTER) (Y/n) message before?

Yes.  I would have thought `add-open = False` would prevent this.  Anyway, sounds like this may be worthy of an issue.  I'll close this conversation as it sounds like we are now on the same page.  Thanks.
